spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/util/GCPLabelUtils.java

```diff
@@ -0,0 +1,223 @@
+  private static Optional<String> getDriverHost(ImmutableMap<String, String> conf) {
+    return Optional.ofNullable(conf.get(SPARK_DRIVER_HOST));
+  }
+
+  /* sample hostname:
+   * sample-cluster-m.us-central1-a.c.hadoop-cloud-dev.google.com.internal */
+  @VisibleForTesting
+  static Optional<String> getClusterName(ImmutableMap<String, String> conf) {
+    return getDriverHost(conf)
+        .map(host -> host.split("\\.")[0])
+        .map(s -> s.contains("-") ? s.substring(0, s.lastIndexOf("-")) : s);
+  }
+
+  @VisibleForTesting
+  static Optional<String> getDataprocRegion(ImmutableMap<String, String> conf) {
+    return fetchGCPMetadata(DATAPROC_REGION_ENDPOINT, conf);
+  }
+
+  @VisibleForTesting
+  static Optional<String> getDataprocJobID(ImmutableMap<String, String> conf) {
+    return getPropertyFromYarnTag(conf, JOB_ID_PREFIX);
+  }
+
+  @VisibleForTesting
+  static Optional<String> getDataprocJobUUID(ImmutableMap<String, String> conf) {
+    return getPropertyFromYarnTag(conf, JOB_UUID_PREFIX);
+  }
+
+  @VisibleForTesting
+  static Optional<String> getDataprocBatchID(ImmutableMap<String, String> conf) {
+    return fetchGCPMetadata(BATCH_ID_ENDPOINT, conf);
+  }
+
+  @VisibleForTesting
+  static Optional<String> getDataprocBatchUUID(ImmutableMap<String, String> conf) {
+    return fetchGCPMetadata(BATCH_UUID_ENDPOINT, conf);
+  }
+
+  @VisibleForTesting
+  static Optional<String> getDataprocSessionID(ImmutableMap<String, String> conf) {
+    return fetchGCPMetadata(SESSION_ID_ENDPOINT, conf);
+  }
+
+  @VisibleForTesting
+  private static Optional<String> getDataprocSessionUUID(ImmutableMap<String, String> conf) {
+    return fetchGCPMetadata(SESSION_UUID_ENDPOINT, conf);
+  }
+
+  @VisibleForTesting
+  static Optional<String> getGCPProjectId(ImmutableMap<String, String> conf) {
+    return fetchGCPMetadata(PROJECT_ID_ENDPOINT, conf)
+        .map(b -> b.substring(b.lastIndexOf('/') + 1));
+  }
+
+  @VisibleForTesting
+  static Optional<String> getSparkAppId(ImmutableMap<String, String> conf) {
+    return Optional.ofNullable(conf.get(SPARK_APP_ID));
+  }
+
+  @VisibleForTesting
+  static Optional<String> getSparkAppName(ImmutableMap<String, String> conf) {
+    return Optional.ofNullable(conf.get(SPARK_APP_NAME));
+  }
+
+  @VisibleForTesting
+  static Optional<String> getClusterUUID(ImmutableMap<String, String> conf) {
+    return fetchGCPMetadata(CLUSTER_UUID_ENDPOINT, conf);
+  }
+
+  @VisibleForTesting
+  static Optional<String> getPropertyFromYarnTag(
+      ImmutableMap<String, String> conf, String tagPrefix) {
+    String yarnTag = conf.get(SPARK_YARN_TAGS);
+    if (yarnTag == null) {
+      return Optional.empty();
+    }
+    return Arrays.stream(yarnTag.split(","))
+        .filter(tag -> tag.startsWith(tagPrefix))
+        .findFirst()
+        .map(tag -> tag.substring(tagPrefix.length()));
+  }
+
+  @VisibleForTesting
+  static Optional<String> fetchGCPMetadata(String httpEndpoint, ImmutableMap<String, String> conf) {
```

Comment:
please cache (using Suppliers.memoize() or regular cache) the output of the metadata server - no point in making multiple calls in the same app

Reply:
made every spark label cached, none of them change in the context of one spark app (at least thats the assumption I'm going with)